app/frontend/terrier/parts/dropzone-part.ts

```diff
@@ -0,0 +1,107 @@
+import {PartTag} from "tuff-core/parts"
+import Messages from "tuff-core/messages"
+import TerrierPart from "./terrier-part"
+import mime from "mime-types"
+
+type DropzonePartState = {
+    file_content?: string | ArrayBuffer
+    dragging_over?: boolean
+    accept_file_types: string[]
+    text?: string
+}
+
+export const FileUploadedKey = Messages.typedKey<{ file: File, file_content: string }>()
+
+export default class DropzonePart extends TerrierPart<DropzonePartState> {
+    _dragOverKey = Messages.untypedKey()
+    _dragLeaveKey = Messages.untypedKey()
+    _dropKey = Messages.untypedKey()
+    _fileInputChangedKey = Messages.untypedKey()
+
+    async init() {
+        this.state.dragging_over = false
+
+        this.onDragOver(this._dragOverKey, m => {
+            m.event.preventDefault()
+            this.state.dragging_over = true
+            this.stale()
+        })
+
+        this.onDragLeave(this._dragLeaveKey, () => {
+            this.state.dragging_over = false
+            this.stale()
+        })
+
+        this.onDrop(this._dropKey, m => {
+            m.event.preventDefault()
+            m.event.stopPropagation()
+            this.state.dragging_over = false
+            this.stale()
+
+            if (m.event.dataTransfer) {
+                if (m.event.dataTransfer.files.length > 1) {
+                    this.alertToast("Please choose only one file to upload.")
+                    return
+                }
```

Comment:
Do we also need a check for `m.event.dataTransfer.files.length > 0`? It's possible someone is dragging structured data that _isn't_ a file, right?